tmail-backend/tmail-third-party/ai-bot/src/main/java/com/linagora/tmail/mailet/rag/httpclient/OpenRagHttpClient.java

```diff
@@ -0,0 +1,157 @@
+ *  useful, but WITHOUT ANY WARRANTY; without even the implied      *
+ *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR         *
+ *  PURPOSE. See the GNU Affero General Public License for          *
+ *  more details.                                                   *
+ ********************************************************************/
+package com.linagora.tmail.mailet.rag.httpclient;
+
+import java.io.ByteArrayInputStream;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Map;
+
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
+import com.linagora.tmail.mailet.rag.RagConfig;
+
+import io.netty.handler.codec.http.HttpMethod;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+
+public class OpenRagHttpClient {
+    private static final Logger LOGGER = LoggerFactory.getLogger(OpenRagHttpClient.class);
+    private static final String API_KEY_HEADER = "Authorization";
+    private static final String CONTENT_TYPE_HEADER = "Content-Type";
+    private static final String APPLICATION_JSON = "application/json";
+
+    private final HttpClient httpClient;
+    private final RagConfig configuration;
+    private final ObjectMapper objectMapper;
+
+    public OpenRagHttpClient(RagConfig configuration) {
+        this.configuration = configuration;
+        this.objectMapper = new ObjectMapper().registerModule(new Jdk8Module());
+        this.httpClient = buildReactorNettyHttpClient(configuration);
+    }
+
+    public Mono<Void> addDocument(Partition partition, DocumentId documentId, String textualContent, Map<String, String> metadata) {
+        String url = String.format("/indexer/partition/%s/file/%s", partition.partitionName(), documentId.asString());
+        return Mono.fromCallable(() -> objectMapper.writeValueAsString(metadata))
+            .flatMap(metadataJson ->
+                sendFormRequest(HttpMethod.POST, url, documentId, textualContent, metadataJson)
+                    .onErrorResume(throwable -> {
+                        if (throwable instanceof DocumentConflictException) {
+                            LOGGER.info("Document already exists. Retrying with PUT.");
+                            return sendFormRequest(HttpMethod.PUT, url, documentId, textualContent, metadataJson);
+                        }
+                        return Mono.error(throwable);
+                    }))
+            .onErrorResume(JsonProcessingException.class, e -> {
+                LOGGER.error("Failed to serialize metadata to JSON", e);
+                return Mono.error(new RuntimeException("Failed to serialize metadata to JSON", e));
+            });
```

Comment:
Log or throw one should choose.

Here I believe that doing nothing is better (Downstream system is unlikely to manage it, and path to recovery is unclear)